src/formatting_stack/formatters/clean_ns.clj

```diff
@@ -57,20 +53,42 @@
 (def default-namespaces-that-should-never-cleaned
   #{'user 'dev})
 
+(defn replaceable-ns-form
+  [{:keys [how-to-ns-opts refactor-nrepl-opts namespaces-that-should-never-cleaned libspec-whitelist]} filename]
+  (when (and (try-require filename)
+             (not (impl/has-duplicate-requires? filename)))
+    (util.ns/replaceable-ns-form
+     filename
+     (make-cleaner how-to-ns-opts refactor-nrepl-opts namespaces-that-should-never-cleaned libspec-whitelist filename)
+     how-to-ns-opts)))
+
 (defn format!
-  [{:keys [how-to-ns-opts refactor-nrepl-opts namespaces-that-should-never-cleaned libspec-whitelist]}
-   files]
+  [this files]
   (->> files
        (process-in-parallel! (fn [filename]
-                               (when (and (try-require filename)
-                                          (not (impl/has-duplicate-requires? filename)))
-                                 (clean! how-to-ns-opts
-                                         refactor-nrepl-opts
-                                         namespaces-that-should-never-cleaned
-                                         libspec-whitelist
-                                         filename)))))
+                               (when-let [ns-replacement (replaceable-ns-form this filename)]
+                                 (println "Cleaning unused imports:" filename)
+                                 (write-ns-replacement! filename ns-replacement)))))
   nil)
 
+(defn lint! [this files]
+  (->> files
+       (process-in-parallel! (fn [filename]
+                               (when-let [{:keys [final-ns-form-str
+                                                  original-ns-form-str]} (replaceable-ns-form this filename)]
+                                 (let [diff (#'cljfmt.diff/unified-diff filename original-ns-form-str final-ns-form-str)]
+                                   (->> (diff->line-numbers diff)
+                                        (mapv (fn [{:keys [begin]}]
```

Comment:
What is your suggestion here? The `diff->line-numbers`-function returns a list of changed sections, for each we need a hashmap. I don't see how we can do that differently

Reply:
Let's put it another way - what do we do if either `diff/unified-diff ` or `diff->line-numbers` throw an exception right here?

It's always possible since Git corner cases are to be expected. Which we wouldn't be particularly prepared against, because:

* `thombergs/diffparser` cannot be considered battle-tested for us (yet)
* `cljfmt` isn't a primarily diff-concerned lib
  * Please imagine what the author's response would be if we found a corner case for something cljfmt does not explicitly support? Or a breaking change?

Needless to say, I'm not against this PR or anything, but enough caution should be placed around the underlying code. Right now it seems the whole linting part revolves around diffs. Seems pretty risky.

Reply:
> Let's put it another way - what do we do if either `diff/unified-diff` or `diff->line-numbers` throw an exception right here?

In both cases it's caught here: https://github.com/nedap/formatting-stack/blob/57e4651fa0913bc4a65415c07b86c3e1dac403a2/src/formatting_stack/util.clj#L99-L103 appended to the result of the report and continues processing.

> `cljfmt` isn't a primarily diff-concerned lib 

All of the formatters are essentially generating a diff and applying that.

Since #163, we _could_ make the diff-parsing opt-in, but it already is. The formatters don't run as linters by default.

I think we need an alpha and some incubation time in some internal projects to see the effects of this PR.

Reply:
> In both cases it's caught here:

Can be good enough, OTOH hitting errors (i.e. an increased error rate) is not desirable, even if those errors are caught and reported.

We can increase confidence around this feature by assessing the `.patch` files that are part of this PR:

**1.-** Do those cover all cases?

* Completetly added file
* Completetly removed file
* Merge conflicts
* Renames
* Other?
  * Worth potentially digging into other projects' test suites

**2.-** Those .patch files would be ideally also exercised against `cljfmt.diff/unified-diff`.

Normally we shouldn't craft tests for _other_ libs, but:

* cljfmt is not a lib that promises to provide an API for diffs
* we're using a defn that is public mostly by chance.
  * So we should create our own wrapper that invokes `cljfmt.diff/unified-diff`
  * WIth this wrapping, it becomes more natural that we craft tests for it.
  * Maybe: vendorize this defn if after asking, we verify that this indeed isn't quite a public API.

> All of the formatters are essentially generating a diff and applying that.

Doesn't seem true to me. Formatters spit contents to filenames. I don't know of a single formatter that emits a .patch (or even: that computes a fine-grained difference at all) and applies it to a file somehow.

The fact that formatting can be _seen_ as patch applications seems orthogonal to the inner workings of these formatters.

(there's a difference between computing diffs after-the-fact, and formatting things out of diffs)

> The formatters don't run as linters by default.

True, but using formatters-as-linters is something various parts will want soonish, so it can be assumed that these will be enabled. And when that happens, builds shouldn't start faiiling. 

> I think we need an alpha and some incubation time in some internal projects to see the effects of this PR.

Agreed, seems quite a precondition to me but also we'd gain a lot of confidence if you could confirm a bit more thoroughtly that we've explored this problem space well, and that we're shipping all reasonable tests, cautions, etc.

Reply:
Thought about this a bit, Will keep the current solution (computing diffs, reporting on the patches).

To increase confidence I extracted the unified-diff function, and added tests for the current usecases. They can be extended whenever we need more.